src/agent.c

```diff
@@ -669,21 +744,52 @@ void agent_run(juice_agent_t *agent) {
 		}
 
 		if (FD_ISSET(agent->sock, &readfds)) {
-			if (agent_recv(agent) < 0)
+			if (agent_recv(agent, (agent_stun_entry_t *)NULL) < 0)
 				break;
 		}
+
+		for (int i = 0; i < agent->entries_count; ++i) {
+			agent_stun_entry_t *entry = agent->entries + i;
+			if (entry->sock == INVALID_SOCKET) continue;
+			if (FD_ISSET(entry->sock, &readfds)) {
+				if (agent_recv(agent, entry) < 0) {
+					entry->state = AGENT_STUN_ENTRY_STATE_FAILED;
+					entry->sock = INVALID_SOCKET;
+				}
+			}
+		}
 	}
 	JLOG_DEBUG("Leaving agent thread");
 	agent_change_state(agent, JUICE_STATE_DISCONNECTED);
 	mutex_unlock(&agent->mutex);
 }
 
-int agent_recv(juice_agent_t *agent) {
+int agent_recv(juice_agent_t *agent, agent_stun_entry_t *entry) {
```

Comment:
There could be a dedicated function to receive from entry transport as the logic looks quite decorelated.